README.md
# DevTools Author [![Build Status](https://travis-ci.org/micjamking/devtools-author.svg?branch=master)](https://travis-ci.org/micjamking/devtools-author)
> A selection of author settings for Chrome Developer Tools.

![Devtools Author Logo](app/images/1400x560_marquee.png)

## Features
If you develop in the browser and/or use Chrome Developer Tools as your web authoring environment, this extension provides a small set of settings to enhance DevTools editing environment:

- Choose from +25 editor themes (applies to [`Elements`, `Sources` and `Console` panels only](https://github.com/micjamking/devtools-author/issues/31))
- Supports system enabled fonts (must be installed prior on your computer).
- Change font size from 10px - 22px

## Usage
1. Install the [Chrome extension](https://chrome.google.com/webstore/detail/devtools-themes/egfhcfdfnajldliefpdoaojgahefjhhi)
2. Enable Developer Tools experiments in `chrome://flags/#enable-devtools-experiments`. Restart Chrome for flags to take effect.
3. Open DevTools (`cmd + opt + I`); *Settings* > *Experiments* > check **Allow custom UI themes**.

After restarting DevTools, you should see the default theme *3024* take affect within the *Sources* and *Console* panels.

## Contributing
[Fork the repository](../../fork) and then follow the steps below to contribute new themes or bug fixes. *Devtools Author* is built using [NodeJS](https://nodejs.org/en/) and [GruntJS](http://gruntjs.com/).

### Installation:
```
$ git clone git@github.com:<username>/devtools-author.git
...
$ cd devtools-author
$ npm install
```

### Development: 
```
$ grunt serve
```
1. Once grunt is running, to install development extension in Chrome, open *Settings* > More Tools > Extensions and click on the **Load unpacked extension...** button. (also enable `Allow incognito` below if you wish).
  - _(Disable DevTools Author if you have the extension installed from the Chrome Web Store.)_
  - Make sure [Developer Tools experiments are enabled and custom UI themes are allowed](#usage).
2. Restart DevTools. I find the fastest way to see changes take affect is to undock DevTools in a separate window and then open a subsequent DevTools window (`cmd + opt + I` while the current DevTools window is focused) after changes have been saved and grunt reloads the assets. You'll then need to reload (close and reopen) the subsequent DevTools window after you make changes.

#### Creating additional themes
1. Make a copy of one of the templates from `app/styles/themes/templates/` and rename the file to your theme name **without the underscore**, ie. if your theme is called *aloha*, inside of `app/styles/themes/`, copy `templates/_theme-template.scss` and rename it to `aloha.scss`
2. Add color values for the palette based on the code syntax highlighter variables in your scss file.
  - If you desire more specific targeting for your theme than what is being supported out of the box, you can add those styles to the end of your theme file, after the `@include styles()`. See [#109](https://github.com/micjamking/devtools-author/issues/109#issuecomment-180145295)
3. Add your color palette object (name and colors array) to the `themes.json` in `app/scripts/`
4. In DevTools, select your theme palette in the *Author Settings* panel.
5. Preview and adjust your colors as needed. See [**Development - Step 2**](#development).
6. Commit and push your changes to your repo, then create a [pull request](../../compare) for your new theme once it is ready!

## Changelog
- [Changelog](CHANGELOG.md)
- [Releases](https://github.com/micjamking/devtools-author/releases)

## License
MIT © Mike King [yo@mikeking.io](mailto:yo@mikeking.io)


CHANGELOG.md
# Change Log

## [v0.6.4](https://github.com/micjamking/devtools-author/tree/v0.6.4) (2016-03-07)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.6.3...v0.6.4)

## [v0.6.3](https://github.com/micjamking/devtools-author/tree/v0.6.3) (2016-02-24)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.6.2...v0.6.3)

**Implemented enhancements:**

- Debugger line in dark themes is illegible [\#130](https://github.com/micjamking/devtools-author/issues/130)
- Improve console.table readability [\#124](https://github.com/micjamking/devtools-author/issues/124)

**Fixed bugs:**

- Debugger line in dark themes is illegible [\#130](https://github.com/micjamking/devtools-author/issues/130)

## [v0.6.2](https://github.com/micjamking/devtools-author/tree/v0.6.2) (2016-02-15)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.6.1...v0.6.2)

**Closed issues:**

- Add 'None/Default' theme option [\#121](https://github.com/micjamking/devtools-author/issues/121)

## [v0.6.1](https://github.com/micjamking/devtools-author/tree/v0.6.1) (2016-02-14)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.6.0...v0.6.1)

## [v0.6.0](https://github.com/micjamking/devtools-author/tree/v0.6.0) (2016-02-07)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.13...v0.6.0)

**Implemented enhancements:**

- Support default dark theme in sources panel in Chrome v50 [\#112](https://github.com/micjamking/devtools-author/issues/112)

## [v0.5.13](https://github.com/micjamking/devtools-author/tree/v0.5.13) (2016-02-07)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.12...v0.5.13)

## [v0.5.12](https://github.com/micjamking/devtools-author/tree/v0.5.12) (2016-02-06)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.11...v0.5.12)

## [v0.5.11](https://github.com/micjamking/devtools-author/tree/v0.5.11) (2016-02-04)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.10...v0.5.11)

**Fixed bugs:**

- cursor is missing on element attribute [\#107](https://github.com/micjamking/devtools-author/issues/107)

## [v0.5.10](https://github.com/micjamking/devtools-author/tree/v0.5.10) (2016-02-04)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.9...v0.5.10)

## [v0.5.9](https://github.com/micjamking/devtools-author/tree/v0.5.9) (2016-02-03)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.8...v0.5.9)

## [v0.5.8](https://github.com/micjamking/devtools-author/tree/v0.5.8) (2016-02-03)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.7...v0.5.8)

**Implemented enhancements:**

- Darcula [\#97](https://github.com/micjamking/devtools-author/issues/97)

## [v0.5.7](https://github.com/micjamking/devtools-author/tree/v0.5.7) (2016-02-03)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.6...v0.5.7)

**Closed issues:**

- Material Dark Theme: Errors not visible :P [\#103](https://github.com/micjamking/devtools-author/issues/103)

## [v0.5.6](https://github.com/micjamking/devtools-author/tree/v0.5.6) (2016-02-02)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.5...v0.5.6)

## [v0.5.5](https://github.com/micjamking/devtools-author/tree/v0.5.5) (2016-02-02)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.4...v0.5.5)

## [v0.5.4](https://github.com/micjamking/devtools-author/tree/v0.5.4) (2016-02-02)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.3...v0.5.4)

## [v0.5.3](https://github.com/micjamking/devtools-author/tree/v0.5.3) (2016-01-31)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.2...v0.5.3)

**Closed issues:**

- Some characters not visible when editing attribute values [\#92](https://github.com/micjamking/devtools-author/issues/92)

## [v0.5.2](https://github.com/micjamking/devtools-author/tree/v0.5.2) (2016-01-31)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.1...v0.5.2)

## [v0.5.1](https://github.com/micjamking/devtools-author/tree/v0.5.1) (2016-01-31)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.5.0...v0.5.1)

## [v0.5.0](https://github.com/micjamking/devtools-author/tree/v0.5.0) (2016-01-29)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.12...v0.5.0)

## [v0.4.12](https://github.com/micjamking/devtools-author/tree/v0.4.12) (2016-01-29)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.11...v0.4.12)

## [v0.4.11](https://github.com/micjamking/devtools-author/tree/v0.4.11) (2016-01-29)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.10...v0.4.11)

## [v0.4.10](https://github.com/micjamking/devtools-author/tree/v0.4.10) (2016-01-29)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.9...v0.4.10)

## [v0.4.9](https://github.com/micjamking/devtools-author/tree/v0.4.9) (2016-01-29)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.8...v0.4.9)

## [v0.4.8](https://github.com/micjamking/devtools-author/tree/v0.4.8) (2016-01-29)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.7...v0.4.8)

## [v0.4.7](https://github.com/micjamking/devtools-author/tree/v0.4.7) (2016-01-28)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.6...v0.4.7)

## [v0.4.6](https://github.com/micjamking/devtools-author/tree/v0.4.6) (2016-01-28)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.5...v0.4.6)

## [v0.4.5](https://github.com/micjamking/devtools-author/tree/v0.4.5) (2016-01-28)
[Full Changelog](https://github.com/micjamking/devtools-author/compare/v0.4.4...v0.4.5)

## [v0.4.4](https://github.com/micjamking/devtools-author/tree/v0.4.4) (2016-01-28)

app/scripts/panel.js
/**
 * @file DevTools Author Settings panel
 */
(function(
  w,        // Window
  ga,       // Google Analytics
  $,        // document.querySelectorAll
  storage   // Chrome Storage API
){
  'use strict';
  
  /**
   * Panel module
   * @namespace panel
   * @global
   */
  var panel = (function(){

    /** @private */
    var $select = $('#theme-options')[0];

    /** @private */
    var $range = $('#font-size-input')[0];

    /** @private */
    var $output = $('#font-size-output')[0];

    /** @private */
    var $fontInput = $('#font-family-input')[0];

    /** @private */
    var $palette = $('.palette')[0];

    /** @private */
    var $themeTitle = $('#currentTheme')[0];

    /** @private */
    var $footerLink = $('.footer a')[0];

    /** @private */
    var $alert = $('.alert')[0];
    
    /** @private */
    var _panel = _panel || {};

    /** @private */
    var _themeJSON = '/dist/scripts/themes.json';

    /** @private */
    var _defaultTheme = '3024';

    /** @private */
    var _defaultFontSize = 14;

    /** @private */
    var _defaultFontFamily = 'Hack';

    /** @private */
    var _currentTheme = null;
   
    /** @private */
    var _currentFontSize = null;
    
    /** @private */
    var _currentFontFamily = null;

    /**
     * Creates HTML list of colors
     * @function _createLI
     * @param {Array} palette - List of theme colors in hex
     */
    function _createLI(palette){
        
      /** Filter palette to remove duplicate colors */
      palette = palette.filter(function(item, pos, self) {
        return self.indexOf(item) === pos;
      });

      for (var i = 0; i < palette.length; i++){
        /** Create list item */
        var item = document.createElement('li');

        /** Style list item */
        item.style.backgroundColor = palette[i];
        item.style.width = (100 / palette.length) + '%';

        /** Add item to the list */
        $palette.appendChild(item);
      }
    
    }

    /**
     * Find theme colors
     * @function _themeLookUp
     * @param {String} theme - Theme name
     * @returns {Array} List of theme colors in hex
     */
    function _themeLookUp(theme){
      
      for (var i = 0; i < _panel.themes.length; i++){
        if (_panel.themes[i].name === theme) {
          return _panel.themes[i].colors;
        }
      }
    
    }

    /**
     * Recreate $palette using current theme
     * @function _updatePalette
     * @param {String} theme - Currently selected theme name
     */
    function _updatePalette(theme){
      var children = $palette.querySelectorAll('li');

      if (children.length){
        for (var i = 0; i < children.length; i++){
          $palette.removeChild(children[i]);
        }
      }

      _createLI(_themeLookUp(theme));
    }

    /**
     * Build select menus like ngOptions
     * @function _buildSelectMenu
     * @param {HTMLElement} menu - Theme select menu element (empty)
     * @param {Object} model - Panel model containing available themes and current settings
     * @returns {HTMLElement} menu - Theme select menu element, with available themes appended
     */
    function _buildSelectMenu(menu, model){
      
      var options, array;

      /** Get the data attribute value */
      options = menu.dataset.options;

      /** Clean string and create array */
      options = options.replace(/in\s/g, '').split(' ');

      /** 
       * Assign array from model by property name
       * using the value from the last item in options 
       */
      array = model[options[options.length - 1]];

      for (var j = 0; j < array.length; j++){

        var option = document.createElement('option');

        /** Assign option value & text from array */
        option.value = array[j].name.replace(/\s+/g, '-').toLowerCase();
        option.text  = array[j].name;

        /** Select currentTheme option */
        if (model.currentTheme === array[j].name){
          option.selected = 'selected';
        }

        menu.add(option, null);
      }

      return menu;
    
    }

    /**
     * Alert notification test
     * @function _notification
     * @param {String} oldValue - Original value for alert notification test
     * @param {String} newValue - New value for alert notification test
     */
    function _notification(oldValue, newValue){
      if (oldValue !== newValue){
        $alert.style.display = 'block';
      }
    }

    /**
     * Object.defineProperty custom setters and getters for _panel model
     * @function _panelModelSetup
     */
    function _panelModelSetup(){

      /** Observe changes to _panel.currentTheme */
      Object.defineProperty(_panel, 'currentTheme', {
        enumerable: true,
        configurable: true,
        get: function() {
          return _currentTheme;
        },
        set: function(newValue) {
          _notification(_currentTheme, newValue);
          $themeTitle.innerHTML = _currentTheme = newValue;
          _updatePalette(_currentTheme);
        }
      });

      /** Observe changes to _panel.currentFontSize */
      Object.defineProperty(_panel, 'currentFontSize', {
        enumerable: true,
        configurable: true,
        get: function() {
          return _currentFontSize;
        },
        set: function(newValue) {
          _notification(_currentFontSize, newValue);
          $range.value = $output.value = _currentFontSize = newValue;
        }
      });

      /** Observe changes to _panel.currentFontFamily */
      Object.defineProperty(_panel, 'currentFontFamily', {
        enumerable: true,
        configurable: true,
        get: function() {
          return _currentFontFamily;
        },
        set: function(newValue) {
          _notification(_currentFontFamily, newValue);
          $fontInput.value = _currentFontFamily = newValue;
        }
      });
    }

    /**
     * Setup event listeners and get settings from storage
     * @function _panelSetup
     */
    function _panelSetup(){
      
      /** Observe changes to _panel model */
      _panelModelSetup();
      
      /** Listen for changes to the select menu */
      $select.addEventListener('change', setTheme);

      /** Listen for changes to the text input */
      $fontInput.addEventListener('change', setFontFamily);

      /** Listen for changes to the range input */
      $range.addEventListener('change', setFontSize);

      /** Listen for click on footer element */
      $footerLink.addEventListener('click', function(){
        ga('send', 'event', 'Link', 'Click', 'Mike King on GitHub');
      });

      /** Get current theme setting from storage */
      storage.get('devtools-theme', function(object){
        _panel.currentTheme = _currentTheme = getTheme( _panel.themes, object['devtools-theme'] );
        _buildSelectMenu($select, _panel);
        
        /** Send Google Analytics 'Install' event on initial install */
        if (!object['devtools-theme'] && _panel.currentTheme === _defaultTheme){
          ga('send', 'event', 'Install', 'Install', 'Devtools Author', 1);
        }
      });

      /** Get current `font-family` setting from storage */
      storage.get('devtools-fontFamily', function(object){
        _panel.currentFontFamily = _currentFontFamily = getFontFamily(object['devtools-fontFamily']);
      });

      /** Get current `font-size` setting from storage */
      storage.get('devtools-fontSize', function(object){
        _panel.currentFontSize = _currentFontSize = getFontSize(object['devtools-fontSize']);
      });
    
    }

    /**
     * Set & save theme based on select menu change event
     * @function setTheme
     * @memberof! app
     * @param {Event} event - Event object
     * @param {Object} obj - Object for theme settings defaults
     */
    function setTheme(event, obj){
      
      function save(theme){
        storage.set({ 'devtools-theme': theme.value }, function(){
          _panel.currentTheme = theme.text;
        });
        return theme.value;
      }
      
      if (event && event.type === 'change'){
        var el     = event.target || event.srcElement;
        var option = el.options[el.selectedIndex];
        return save(option);
      } else if (event === null && obj){
        return save(obj);
      }
    
    }

    /**
     * Set & save `font-family` based on input change event
     * @function setFontFamily
     * @memberof! app
     * @param {Event} event - Event object
     * @param {Object} value - Value for `font-family` settings
     */
    function setFontFamily(event, value){
      
      function save(fontFamily){
        storage.set({ 'devtools-fontFamily': fontFamily }, function(){
          _panel.currentFontFamily = fontFamily;
        });
        return fontFamily;
      }
      
      if (event && event.type === 'change'){
        var el = event.target || event.srcElement;
        return save(el.value);
      } else if (event === null && value){
        return save(value);
      }
    
    }

    /**
     * Set & save `font-size` based on input menu change event
     * @function setFontSize
     * @memberof! app
     * @param {Event} event - Event object
     * @param {Object} value - Value for `font-size` settings
     */
    function setFontSize(event, value){
      
      function save(fontSize){
        storage.set({ 'devtools-fontSize': fontSize }, function(){
          _panel.currentFontSize = fontSize;
        });
        return fontSize;
      }
      
      if (event && event.type === 'change'){
        var el = event.target || event.srcElement;
        return save(el.value);
      } else if (event === null && value){
        return save(value);
      }
    
    }

    /**
     * Get theme settings
     * @function getTheme
     * @memberof! app
     * @param {Array} array - List of themes
     * @param {String} string - Value for theme settings
     * @returns {String} Theme name
     */
    function getTheme(array, string){
      
      if (!array || !string){
        setTheme(null, {
          value: _defaultTheme.replace(/\s+/g, '-').toLowerCase(),
          text: _defaultTheme
        });
        return _defaultTheme;
      }
      
      for (var i = 0; i < array.length; i++){
        if (array[i].name.replace(/\s+/g, '-').toLowerCase() === string){
          return array[i].name;
        }
      }
    
    }

    /**
     * Get `font-family` settings
     * @function getFontFamily
     * @memberof! app
     * @param {String} value - Value for `font-family` settings
     * @returns {String} value - Value for `font-family` settings
     */
    function getFontFamily(value){
      
      if (!value){
        setFontFamily(null, _defaultFontFamily);
        return _defaultFontFamily;
      } else {
        return value;
      }
    
    }

    /**
     * Get `font-size` settings
     * @function getFontSize
     * @memberof! app
     * @param {String} value - Value for `font-size` settings
     * @returns {String} value - Value for `font-size` settings
     */
    function getFontSize(value){
      
      if (!value){
        setFontSize(null, _defaultFontSize);
        return _defaultFontSize;
      } else {
        return value;
      }
    
    }

    /**
     * GET themes JSON and setup panel settings
     * @function init
     * @memberof! app
     */
    function init(){

      var ajax = new XMLHttpRequest();
      
      ajax.open('GET', _themeJSON);
      ajax.send(null);
      
      ajax.onreadystatechange = function(){
        if (ajax.readyState === 4) {
          if (ajax.status === 200) {
            _panel.themes = JSON.parse(ajax.responseText);

            _panel.themes.sort(function(a, b){
              if(a.name < b.name) { return -1; }
              if(a.name > b.name) { return 1; }
              return 0;
            });

            return _panelSetup();
          } else {
            return console.log('Status Code: ' + ajax.status + '\nThere was an error with your request');
          }
        }
      };
    
    }

    /** Public methods */
    return {
      init: init,
      getTheme: getTheme,
      setTheme: setTheme,
      getFontSize: getFontSize,
      setFontSize: setFontSize,
      getFontFamily: getFontFamily,
      setFontFamily: setFontFamily
    };
  })();
  
  /** Initialize panel */
  panel.init();

  /** Export as global */
  w.panel = panel;

})(
  /** Globals */
  window,
  window.ga,
  document.querySelectorAll.bind(document),
  chrome.storage.sync
);